content/doc/book/pipeline/jenkinsfile.adoc

```diff
@@ -889,6 +854,97 @@ within a single `withCredentials( ... ) { ... }` step by doing the following:
 
 
+=== String interpolation
+
+Jenkins Pipeline uses rules identical to link:http://groovy-lang.org[Groovy] for
+string interpolation. Groovy's String interpolation support can be confusing to
+many newcomers to the language. While Groovy supports declaring a string with
+either single quotes, or double quotes, for example:
+
+[source,groovy]
+----
+def singlyQuoted = 'Hello'
+def doublyQuoted = "World"
+----
+
+Only the latter string will support the dollar-sign (`$`) based string
+interpolation, for example:
+
+[source,groovy]
+----
+def username = 'Jenkins'
+echo 'Hello Mr. ${username}'
+echo "I said, Hello Mr. ${username}"
+----
+
+Would result in:
+
+[source]
+----
+Hello Mr. ${username}
+I said, Hello Mr. Jenkins
+----
+
+Understanding how to use string interpolation is vital for using some of
+Pipeline's more advanced features.
+
+[WARNING]
+.Interpolation of sensitive environment variables
+====
+
+Groovy string interpolation should [red]*never* be used with sensitive environment variables (i.e. credentials, see: <<Handling credentials>>).
+This is because the sensitive environment variable will be interpolated at Groovy evaluation, and the environment variable's value could be made available earlier than intended, resulting in the sensitive data leaking.
+
+For example, consider a sensitive environment variable passed to the shell step.
+Should Groovy perform the interpolation, the sensitive value would become available not only in the shell step, but the Jenkins Controller as well.
+```
+node {
+  withCredentials([usernamePassword(credentialsId: 'creds', usernameVariable: 'USERNAME', passwordVariable: 'PASSWORD')] {
+    /* WRONG! */
+    sh "curl -u ${USERNAME}:${PASSWORD} https://example.com/"
+    /* CORRECT */
+    sh 'curl -u $USERNAME:$PASSWORD https://example.com/'
+  }
+}
+```
+
+====
+
+[WARNING]
+.Injection via interpolation
+====
+
+Another note of caution. With Groovy string interpolation, it is possible to perform SQL-like injections.
+This occurs when a variable (any type, not just environment variables) that contains special characters (e.g. `' " / \ $ & % ^ > < | ;`) is passed to the `sh`, `bat`, `powershell`, or `pwsh` steps.
+For a simple example:
+```
+node {
+  withCredentials([string(credentialsId: 'sec-ret-with-symbols', variable: 'SECRET_VALUE')]) {
+    // value of credential = sec>ret
+    sh "echo $SECRET_VALUE"
+  }
+}
+```
+In this example, the string `"echo ${SECRET_VALUE"` gets interpolated by Groovy, and the value `echo sec>ret` gets sent to the shell step.
+In the shell nomenclature, this does not echo the value `sec>ret`, but instead will create a file called `ret` that will contain the value `sec`.
+It is easy to see how a carefully constructed variable could perform any number of undesired shell commands.
+
+Another unintended consequence of these accidental injections is that it can slightly mangle credential values such that password masking fails in the console log.
```

Comment:
```suggestion
This kind of issue also sometimes occurs accidentally when credentials that contain special characters are passed to a step using Groovy interpolation.
This can result in the credentials being mangled, causing them to be invalid, and preventing them from being automatically masked in the console log.
```

Reply:
resolved in de21b90b62357588df4ac3af80df66263a0c6e7a